### PR TITLE
Add Patcher manual patch. Add patch for Reversible Tram AI mod

### PR DESCRIPTION
**Pull request by sway2020**
Fixes: #1361 

Related issue: #1353 

I forced pushed to the master branch so I could no longer re-open my old PR, so I sent a new PR instead. I tested it and it should work with the current workshop version of the reversible tram ai mod (v0.3.1)

https://steamcommunity.com/sharedfiles/filedetails/?id=2740907672

### CODE REVIEW

File: TLM/TLM/Lifecycle/Patcher.cs
Comment (aubergine10) [inline TLM/TLM/Lifecycle/Patcher.cs:48]: Would this be more legible as:

```csharp
fail |= !PatchManual(HARMONY_ID);
```
?

File: TLM/TLM/Lifecycle/Patcher.cs
Comment (aubergine10) [inline TLM/TLM/Lifecycle/Patcher.cs:124]: What happens if the assembly is enabled _after_ this code runs.

For example, user loads game to main menu, then in Steam they subscribe Reversible Tram AI mod (or they might already have it subscribed, but disabled) , then they enable it....?

I suspect we'll need to listen for plugin manager events (hooked via `TMPELifecycle`) to invoke `PatchManual()` whenever a mod is enabled/disabled, and in turn `PatchManual()` will need to test if it's already run, or something along those lines?

**Review by aubergine10 — CHANGES_REQUESTED** [review]

File: TLM/TLM/Lifecycle/Patcher.cs
Comment (aubergine10) [reply TLM/TLM/Lifecycle/Patcher.cs:124]: Additionally, once patched, what happens if the user disables Reversible Tram AI mod? Do we need to also remove patches that TM:PE added?

File: TLM/TLM/Lifecycle/Patcher.cs
Comment (sway2020) [reply TLM/TLM/Lifecycle/Patcher.cs:48]: This one is just personal preference. We can change it to match TMPE'S style

Comment (aubergine10) [reply TLM/TLM/Lifecycle/Patcher.cs:48]: It's a huge codebase - every little bit helps :)

File: TLM/TLM/Lifecycle/Patcher.cs
Comment (sway2020) [reply TLM/TLM/Lifecycle/Patcher.cs:124]: I am not sure how TMPE uses the Patcher class. My mods usually apply manual patches in OnLevelLoaded(). If users disable the reversible tram ai mod, Harmony patches in the ai mod will be unpatched. The patch in this PR is a patch for another patch, so I think TMPE doesn't need to remove this patch if the reversible tram ai mod is disabled

Comment (aubergine10) [reply TLM/TLM/Lifecycle/Patcher.cs:124]: > The patch in this PR is a patch for another patch, so I think TMPE doesn't need to remove this patch if the reversible tram ai mod is disabled

What if user disables TM:PE after it's patched your patch? Example: User is trying to track down a crash or some other bug, suspects it's TM:PE so disables TM:PE (which has already patched your patch) then loads city.

Comment (aubergine10) [reply TLM/TLM/Lifecycle/Patcher.cs:124]: @krzychu124 any thoughts on this? Is it something that needs to go in this PR or can it be added at later date? It would be good to get this in for 11.6.5 release if possible, assuming it's going to TEST workshop page first as that way we can start getting feedback from users.

Comment (sway2020) [reply TLM/TLM/Lifecycle/Patcher.cs:124]: > What if user disables TM:PE after it's patched your patch? Example: User is trying to track down a crash or some other bug, suspects it's TM:PE so disables TM:PE (which has already patched your patch) then loads city.

I will test this later

Comment (krzychu124) [reply TLM/TLM/Lifecycle/Patcher.cs:124]: Lifecycle of patches was completely redesigned recently.

Patcher runs first time (only Pathfinding patches) when you click `Load Game` or start `New game` (level preload), because game is "warming up" pathfinding to fix paths of loaded public transport lines and `OnLevelLoaded` is way too late to do so - you would see disconnected lines after load, then runs again but this time with all remaining patches at the end of loading process (`OnLevelLoaded`).

> What if user disables TM:PE after it's patched your patch?

Obviously all TMPE patches are reverted (unpatched) when you disable it, no issue here.

> Additionally, once patched, what happens if the user disables Reversible Tram AI mod? Do we need to also remove patches that TM:PE added?

Same here, we are patching "Reversible Tram AI" - it's AI that drives trams -> no mod, no AI (vehicles should despawn but may throw error when assigned to line (not tested that)) - nothing will trigger that patch.


IMO it's ok, code looks good. I've tested it multiple times + few hot-reloads. Didn't notice any issues besides some weird situations which problably require a lot more complex code to solve (arriving trams from both sides of tram stop segment sometimes cause derail if connecting segments are not straight XD) 
We can release it in 11.6.5, I don't see anything that is holding it longer.

Comment (sway2020) [reply TLM/TLM/Lifecycle/Patcher.cs:124]: Yes there may still be some weird edge cases like the one you mentioned. And I'm a bit worried that users might mistakenly think it's caused by TMPE and bother the TMPE team

**Review by krzychu124 — APPROVED** [review]
Looks ok, tested multiple times in game, no visible issues.

File: TLM/TLM/Lifecycle/Patcher.cs
Comment (sway2020) [reply TLM/TLM/Lifecycle/Patcher.cs]: Fixed in https://github.com/CitiesSkylinesMods/TMPE/pull/1386/commits/558c9a8bb370d5c45046e0ff1c95a3ab7d9cfa60

**Review by aubergine10 — APPROVED** [review]
LGTM 👍

**Comment by aubergine10** [comment]
@krzychu124 does sway have privs to merge this PR to `master`?

**Comment by krzychu124** [comment]
No idea, but ping me @sway2020 if ready :) I'll merge if you don't have privs.

**Comment by sway2020** [comment]
GitHub says I don't have write access. I don't think I will ever need to change anything in the patch target method in my mod (SimulationStepPatch1) so I think this PR is ready to be merged

**Pull request merged**
